File: 2020/3/solve_p1.hs
#!/usr/bin/env runhaskell

{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as T
import qualified Data.Text.IO as TIO

moveTo :: Int -> (Int, Int)
moveTo y = (3 * y `mod` 31, y)

navigateTo :: Int -> [(Int, Int)]
navigateTo targetY = [moveTo y | y <- [1..targetY]]

main :: IO ()
main = do
  world <- T.lines <$> TIO.readFile "in.txt"

  let points = navigateTo (length world - 1)
      traversedTiles = map (\(x, y) -> T.index (world !! y) x) points
      solution = (length . filter (== '#')) traversedTiles

  print solution

File: 2020/3/solve_p1.py
#!/usr/bin/env python3

def move_to(y):
    return (3 * y % 31, y)


def navigate_to(y):
    return [move_to(y) for y in range(1, y)]


def main():
    world = open('in.txt', 'r').readlines()
    world = [x.strip() for x in world]

    points = navigate_to(len(world))
    traversed_tiles = map(lambda p: world[p[1]][p[0]], points)
    solution = sum(1 for tile in traversed_tiles if tile == '#')

    print(solution)


if __name__ == '__main__':
    main()
